File: frontend/app/services/roleplayService.ts
import { apiRequest } from "./apiClient";
import type { ApiEnvelope } from "../state/types";

export async function createRoleplaySession(payload: {
  scenario_id: string;
  level: string;
  display_preferences: Record<string, unknown> | null;
}): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: "/api/v1/roleplay/sessions",
    auth: true,
    body: payload,
  });
}

export async function submitRoleplayTurn(sessionId: string, payload: { user_message: string }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/roleplay/sessions/${sessionId}/turns`,
    auth: true,
    body: payload,
  });
}

export async function fetchRoleplaySession(sessionId: string): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/roleplay/sessions/${sessionId}`,
    auth: true,
  });
}

export async function fetchRoleplayTranscript(sessionId: string): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/roleplay/sessions/${sessionId}/transcript`,
    auth: true,
  });
}

export async function fetchRoleplayReview(sessionId: string): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/roleplay/sessions/${sessionId}/review`,
    auth: true,
  });
}

File: backend/services/roleplay_service.py
from __future__ import annotations

from typing import Any

from ..roleplay.runtime import (
    create_roleplay_session as create_roleplay_session_logic,
    get_roleplay_review as get_roleplay_review_logic,
    get_roleplay_session as get_roleplay_session_logic,
    get_roleplay_transcript as get_roleplay_transcript_logic,
    submit_roleplay_turn as submit_roleplay_turn_logic,
)


def create_roleplay_session(*, user_id: str, scenario_id: str, level: str, display_preferences: dict[str, Any] | None) -> dict[str, Any]:
    return create_roleplay_session_logic(user_id=user_id, scenario_id=scenario_id, level=level, display_preferences=display_preferences)


def submit_roleplay_turn(*, user_id: str, session_id: str, user_message: str) -> dict[str, Any]:
    return submit_roleplay_turn_logic(user_id=user_id, session_id=session_id, user_message=user_message)


def get_roleplay_session(*, user_id: str, session_id: str) -> dict[str, Any]:
    return get_roleplay_session_logic(user_id=user_id, session_id=session_id)


def get_roleplay_transcript(*, user_id: str, session_id: str) -> dict[str, Any]:
    return get_roleplay_transcript_logic(user_id=user_id, session_id=session_id)


def get_roleplay_review(*, user_id: str, session_id: str) -> dict[str, Any]:
    return get_roleplay_review_logic(user_id=user_id, session_id=session_id)
